### DISCUSSION TITLE
Plugins/extensions using filesystem paths and the `-f` option

### DISCUSSION THREAD

**Discussion by pawamoy** [mkdocs/mkdocs#2833] [General]
A bit of context: there are some plugins / markdown extensions that use the file system some way or another.
They provide configurable options, for example `base_path` in `pymdownx.snippets` or `watch` (and soon `path`) in `mkdocstrings` and handlers.
Usually, these plugins and extensions will use these paths as relative to the current working directory.

The issue is, this current working directory is not always the same. For example some users try to build other projects' docs, to integrate several sites into a single one. They use the `mkdocs build -f path/to/config.yml` command from another directory.

As soon as they do this, every plugin/extension that depend on the current working directory stop working.
Plugins at least can know the location of the configuration file, to compute their paths relative to it, so that the current working directory has no impact anymore. But Markdown extensions are not supposed to know anything about MkDocs or its config file.

Examples:

- https://github.com/mkdocstrings/mkdocstrings/issues/311
- https://github.com/oprypin/mkdocs-gen-files/issues/5
- (give me links and I'll update)

So I wonder: would there be some kind of generic way to ensure that every plugin/extension that use filesystem paths can work without depending on the current working directory?

**Comment by pawamoy** [comment]
Maybe there should be a configurable, global option in Python-Markdown itself, to set the base directory for relative paths, and MkDocs would pass it the docs directory. Extensions would then be able to fetch that value using Python-Markdown's public API. Or the value could be fed to every extension as well. Or more directly, Python-Markdown would provide a `resolve(relpath)` method that extensions can use if they want to avoid depending on the current working directory.

This would effectively serve as a bridge between SSGs (or other tools using Python-Markdown) and Python-Markdown extensions themselves. Each tool would be responsible for setting the correct / a sensible base directory.

**Comment by oprypin** [comment]
Related:

* https://github.com/mkdocs/mkdocs/issues/2154
* https://github.com/oprypin/mkdocs-gen-files/issues/5
* https://github.com/mkdocs/mkdocs/pull/2642#discussion_r737862344

**Reply by pawamoy** [reply]
Thanks, I've added the gen-files issue to the list.

**Reply by pawamoy** [reply]
And in fact the idea mentioned in the MkDocs PR you linked seems to be a nice candidate. At least that would solve the issue for plugins: MkDocs provides a `ListOfPaths` type that plugin writers can use so that paths given by users are automatically computed as relative to mkdocs.yml.

Now I'm not sure how we could bring the same thing for Markdown extensions, as MkDocs has no way  to know which option is a relative path or not.

**Comment by oprypin** [comment]
For plugins actually we already have great options available

https://github.com/mkdocs/mkdocs/blob/0437b7f85380a409244e3b4fabcfb84dccdf6efa/mkdocs/config/config_options.py#L366

https://github.com/mkdocs/mkdocs/blob/0437b7f85380a409244e3b4fabcfb84dccdf6efa/mkdocs/config/config_options.py#L424


For Markdown extensions there's no answer yet
